sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala

```diff
@@ -358,21 +518,172 @@ private case class GpuParquetFileFilterHandler(@transient sqlConf: SQLConf) exte
+        inputStream.seek(footerLengthIndex)
+        val footerLength = readIntLittleEndian(inputStream)
+        val magic = new Array[Byte](MAGIC.length)
+        inputStream.readFully(magic)
+        if (!util.Arrays.equals(MAGIC, magic)) {
+          throw new RuntimeException(s"$filePath is not a Parquet file. " +
+              s"Expected magic number at tail ${util.Arrays.toString(MAGIC)} " +
+              s"but found ${util.Arrays.toString(magic)}")
+        }
+        val footerIndex = footerLengthIndex - footerLength
+        if (footerIndex < MAGIC.length || footerIndex >= footerLengthIndex) {
+          throw new RuntimeException(s"corrupted file: the footer index is not within " +
+              s"the file: $footerIndex")
+        }
+        inputStream.seek(footerIndex)
+        closeOnExcept(HostMemoryBuffer.allocate(footerLength, false)) { outBuffer =>
+          val out = new HostMemoryOutputStream(outBuffer)
+          val tmpBuffer = new Array[Byte](4096)
+          var bytesLeft = footerLength
+          while (bytesLeft > 0) {
+            val readLength = Math.min(bytesLeft, tmpBuffer.length)
+            inputStream.readFully(tmpBuffer, 0, readLength)
+            out.write(tmpBuffer, 0, readLength)
+            bytesLeft -= readLength
+          }
+          outBuffer
+        }
+      }
+    }
+    withResource(footerBuffer) { footerBuffer =>
+      withResource(new NvtxRange("Parse and filter footer by range", NvtxColor.RED)) { _ =>
+        val len = if (fileLen <= file.length) {
+          // secret signal to skip filtering
+          -1
+        } else {
+          file.length
+        }
+        ParquetFooter.readAndFilter(footerBuffer, file.start, len,
+          names, children, readDataSchema.length, !isCaseSensitive)
+      }
+    }
+  }
+
+  def readAndSimpleFilterFooter(
+      file: PartitionedFile,
+      conf : Configuration,
+      filePath: Path): ParquetMetadata = {
+    //noinspection ScalaDeprecation
+    withResource(new NvtxRange("readFooter", NvtxColor.YELLOW)) { _ =>
+      ParquetFileReader.readFooter(conf, filePath,
+        ParquetMetadataConverter.range(file.start, file.start + file.length))
+    }
+  }
+
+  @scala.annotation.nowarn
   def filterBlocks(
+      footerReader: ParquetFooterReaderType.Value,
       file: PartitionedFile,
       conf : Configuration,
       filters: Array[Filter],
       readDataSchema: StructType): ParquetFileInfoWithBlockMeta = {
     withResource(new NvtxRange("filterBlocks", NvtxColor.PURPLE)) { _ =>
       val filePath = new Path(new URI(file.filePath))
-      //noinspection ScalaDeprecation
-      val footer = withResource(new NvtxRange("readFooter", NvtxColor.YELLOW)) { _ =>
-        ParquetFileReader.readFooter(conf, filePath,
-          ParquetMetadataConverter.range(file.start, file.start + file.length))
+      val footer = footerReader match {
+        case ParquetFooterReaderType.NATIVE =>
+          System.err.println("NATIVE FOOTER READER...")
```

Comment:
Switch to real logging once done with WIP

Reply:
agree with @gerashegalov